src/document-context/measure-custom-elements.ts

```diff
@@ -0,0 +1,382 @@
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+/** options passed to document.registerElement */
+interface RegisterElementOptions {
+  prototype?: {
+    createdCallback?: Function;
+
+    attachedCallback?: Function;
+    detachedCallback?: Function;
+
+    attributeChangedCallback?: Function;
+
+    _propertySetter?: Function;
+    notifyPath?: Function;
+  };
+  extends?: string;
+}
+type CustomElementV0Constructor = {
+  constructor(): HTMLElement
+};
+interface Document {
+  registerElement?(tagName: string, options: RegisterElementOptions):
+      CustomElementV0Constructor;
+}
+interface PolymerSubset {
+  (): void;
+  Async?: {
+    //
+    run(callback: Function, waitTime: number): void;
+  };
+  RenderStatus?: {
+    //
+    whenReady(cb: Function): void;
+  };
+}
+interface Window {
+  _printElementStats(): void;
+  _getElementMeasures(): ElementMeasurement[];
+  _summarizeRange(start: number, end: number): void;
+
+  customElements?: CustomElementsRegistry;
+}
+
+interface CustomElementsRegistry {
+  define(tagName: string, constructor: CustomElementV1Constructor):
+      CustomElementV1Constructor;
+}
+
+interface CustomElementV1Constructor {
+  new (): V1CustomElement;
+}
+
+interface V1CustomElement extends HTMLElement {
+  connectedCallback?(): void;
+  disconnectedCallback?(): void;
+  attributeChangedCallback?(): void;
+}
+
+interface ElementMeasurement {
+  tagName: string;
+  operation: string;
+  elementId: number|undefined;
+  duration: number;
+  start: number;
+  end: number;
+}
+
+type CallbackName = 'register' | 'created' | 'attached' | 'detached' |
+    'attributeChanged' | 'data';
+
+interface Console {
+  timeStamp(label: string): void;
+}
+
+(() => {
+  'use strict';
+
+  // No point without the measurement APIs we need (or polyfills thereof).
+  if (!window.performance || !window.performance.mark ||
+      !window.performance.measure) {
+    return;
+  }
+
+  const prefix = '🐰💫 ';
+
+  function makeMeasurement<R>(
+      operation: string, tagName: string, counter: number | null,
+      cb: () => R): R {
+    const counterSuffix = counter == null ? '' : ` ${counter}`;
+    const startMark = `${prefix}start ${operation} ${tagName}${counterSuffix}`;
+    const endMark = `${prefix}end ${operation} ${tagName}${counterSuffix}`;
+    const measure = `${prefix}${operation} ${tagName}${counterSuffix}`;
+
+    window.performance.mark(
+        `${prefix}start ${operation} ${tagName}${counterSuffix}`);
+    try {
+      return cb();
+    } finally {
+      window.performance.mark(endMark);
+      window.performance.measure(measure, startMark, endMark);
+    }
+  };
+
+  const idSymbol = Symbol('id');
+
+  const originalRegisterElement = document.registerElement;
+  if (originalRegisterElement) {
+    const boundRegisterElement: typeof originalRegisterElement =
```

Comment:
typescript doesn't understand `Function#bind` well at all. This is just saying that `boundRegisterElement` has the same type as `originalRegisterElement`.